test/server/guarddog_impl_test.cc

```diff
@@ -0,0 +1,156 @@
+ * green in the coverage report. However, rest assured from the results of the
+ * test: these lines are in fact covered.
+ */
+class GuardDogDeathTest : public testing::Test {
+protected:
+  GuardDogDeathTest()
+      : config_kill_(1000, 1000, 100, 1000), config_multikill_(1000, 1000, 1000, 500),
+        time_point_(std::chrono::system_clock::now()) {}
+
+  NiceMock<Configuration::MockMain> config_kill_;
+  NiceMock<Configuration::MockMain> config_multikill_;
+  NiceMock<Stats::MockStore> fakestats_;
+  MockSystemTimeSource time_source_;
+  std::chrono::system_clock::time_point time_point_;
+};
+
+class GuardDogMissTest : public testing::Test {
+protected:
+  GuardDogMissTest() : config_miss(1, 1000, 0, 0), config_mega(1000, 1, 0, 0) {}
+
+  NiceMock<Configuration::MockMain> config_miss;
+  NiceMock<Configuration::MockMain> config_mega;
+  NiceMock<Stats::MockStore> stats_store;
+  MockSystemTimeSource time_source_;
+  std::chrono::system_clock::time_point time_point_;
+};
+
+TEST_F(GuardDogDeathTest, KillDeathTest) {
+  // Is it German for "The Function"? Almost...
+  auto die_function = [&]() -> void {
+    InSequence s;
+    EXPECT_CALL(time_source_, currentSystemTime()).WillRepeatedly(testing::Return(time_point_));
+    GuardDogImpl gd(fakestats_, config_kill_, time_source_);
+    auto unpet_dog = gd.createWatchDog(0);
+    gd.forceCheckForTest();
+    time_point_ += std::chrono::milliseconds(500);
+    EXPECT_CALL(time_source_, currentSystemTime()).WillRepeatedly(testing::Return(time_point_));
+    gd.forceCheckForTest();
+  };
+  // Why do it this way? Any threads must be started inside the death test
+  // statement and this is the easiest way to accomplish that.
+  EXPECT_DEATH(die_function(), "");
+}
+
+TEST_F(GuardDogDeathTest, MultiKillDeathTest) {
+  auto die_function = [&]() -> void {
+    InSequence s;
+    EXPECT_CALL(time_source_, currentSystemTime()).WillRepeatedly(testing::Return(time_point_));
+    GuardDogImpl gd(fakestats_, config_multikill_, time_source_);
+    auto unpet_dog1 = gd.createWatchDog(0);
+    gd.forceCheckForTest();
+    auto unpet_dog2 = gd.createWatchDog(1);
+    gd.forceCheckForTest();
+    time_point_ += std::chrono::milliseconds(501);
+    EXPECT_CALL(time_source_, currentSystemTime()).WillRepeatedly(testing::Return(time_point_));
+    gd.forceCheckForTest();
+  };
+  EXPECT_DEATH(die_function(), "");
+}
+
+TEST_F(GuardDogDeathTest, NearDeathTest) {
+  InSequence s;
+  EXPECT_CALL(time_source_, currentSystemTime()).WillRepeatedly(testing::Return(time_point_));
+  GuardDogImpl gd(fakestats_, config_multikill_, time_source_);
+  auto unpet_dog = gd.createWatchDog(0);
+  auto pet_dog = gd.createWatchDog(1);
+  for (int i = 0; i < 6; i++) {
+    time_point_ += std::chrono::milliseconds(100);
+    EXPECT_CALL(time_source_, currentSystemTime()).WillRepeatedly(testing::Return(time_point_));
+    pet_dog->touch();
+    gd.forceCheckForTest();
+  }
+}
+
+TEST_F(GuardDogMissTest, MissTest) {
```

Comment:
It might be helpful to the reader to have a short english explanation of what the test does above. E.g. "This test validates that when we miss that stats are incremented."